src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.Websockets.cs

```diff
@@ -0,0 +1,235 @@
+                                        (IntPtr)_thisHandle,
+                                        out dwReceivedBytes,
+                                        out fCompletionExpected);
+            if (!fCompletionExpected)
+            {
+                CompleteReadWebSockets(hr, dwReceivedBytes);
+            }
+
+            return _readWebSocketsOperation;
+        }
+
+        private unsafe IISAwaitable WriteWebSocketsFromIISAsync(ReadOnlyBuffer<byte> buffer)
+        {
+            var fCompletionExpected = false;
+            var hr = 0;
+            var nChunks = 0;
+
+            if (buffer.IsSingleSegment)
+            {
+                nChunks = 1;
+            }
+            else
+            {
+                foreach (var memory in buffer)
+                {
+                    nChunks++;
+                }
+            }
+
+            if (buffer.IsSingleSegment)
+            {
+                var pDataChunks = stackalloc HttpApiTypes.HTTP_DATA_CHUNK[1];
+
+                fixed (byte* pBuffer = &MemoryMarshal.GetReference(buffer.First.Span))
+                {
+                    ref var chunk = ref pDataChunks[0];
+
+                    chunk.DataChunkType = HttpApiTypes.HTTP_DATA_CHUNK_TYPE.HttpDataChunkFromMemory;
+                    chunk.fromMemory.pBuffer = (IntPtr)pBuffer;
+                    chunk.fromMemory.BufferLength = (uint)buffer.Length;
+                    hr = NativeMethods.http_websockets_write_bytes(_pInProcessHandler, pDataChunks, nChunks, IISAwaitable.WriteCallback, (IntPtr)_thisHandle, out fCompletionExpected);
+                }
+            }
+            else
+            {
+                // REVIEW: Do we need to guard against this getting too big? It seems unlikely that we'd have more than say 10 chunks in real life
+                var pDataChunks = stackalloc HttpApiTypes.HTTP_DATA_CHUNK[nChunks];
+                var currentChunk = 0;
+
+                // REVIEW: We don't really need this list since the memory is already pinned with the default pool,
+                // but shouldn't assume the pool implementation right now. Unfortunately, this causes a heap allocation...
+                var handles = new MemoryHandle[nChunks];
+
+                foreach (var b in buffer)
+                {
+                    ref var handle = ref handles[currentChunk];
+                    ref var chunk = ref pDataChunks[currentChunk];
+
+                    handle = b.Retain(true);
+
+                    chunk.DataChunkType = HttpApiTypes.HTTP_DATA_CHUNK_TYPE.HttpDataChunkFromMemory;
+                    chunk.fromMemory.BufferLength = (uint)b.Length;
+                    chunk.fromMemory.pBuffer = (IntPtr)handle.Pointer;
+
+                    currentChunk++;
+                }
+                hr = NativeMethods.http_websockets_write_bytes(_pInProcessHandler, pDataChunks, nChunks, IISAwaitable.WriteCallback, (IntPtr)_thisHandle, out fCompletionExpected);
```

Comment:
Where is this hr checked?

Reply:
Forgot the synchronous completion check.